### PR TITLE
Added support for nulls in collections

### PR DESCRIPTION
**Pull request by zsperske**
This adds support for null values inside arrays, see the test files for an example. I don't believe this would work for a null value in an array of arrays though. Should support for that be added at the same time? Also what other tests should be added? Should a test be added for updating values in an array like this?

### CODE REVIEW

**Review by traviswestbrook — CHANGES_REQUESTED** [review]
Do you think its worth adding a test for just `[null]`

**Comment by zsperske** [comment]
> Do you think its worth adding a test for just [null]

I started with that, but yeah probably would be good to have a separate test.

**Comment by KennethNickles** [comment]
Do we test for maps/dictionaries with null values?

**Comment by zsperske** [comment]
> Do we test for maps/dictionaries with null values?

No, should we add support for that possibility? Should probably be in another PR.

**Review by KennethNickles — APPROVED** [review]

File: demo/src/test/java/com/workday/autoparse/json/demo/InstanceUpdaterTest.java [hunk 1]
```diff
@@ -353,6 +358,19 @@ public void testJSONArray() throws Exception {
...
+    public void testCollectionWithNullValueUpdate() throws Exception {
+        TestObject testObject = new TestObject();
+        testObject.myCollectionWithNullValues = Arrays.asList("test", null, "null");;
+
+        Map<String, Object> updates = getUpdateMapFromFile("update-json-array-with-nulls.json");
+        
+        TestObject$$JsonObjectParser.INSTANCE.updateInstanceFromMap(testObject, updates, CONTEXT);
+
+        Collection<String> expected = Arrays.asList("one", null, "null");
+        Assert.assertThat(testObject.myCollectionWithNullValues, CoreMatchers.is(expected));
```
Comment (ndtaylor) [inline demo/src/test/java/com/workday/autoparse/json/demo/InstanceUpdaterTest.java]: Use `assertEquals`. This style isn't used anywhere else in the project.

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java [hunk 2]
```diff
@@ -254,6 +260,15 @@ private void testParse(String fileName) throws Exception {
...
         assertEquals("testObject.myDefaultCollection", Collections.singleton("the one"),
                      testObject.myDefaultCollection);
+        final Collection<String> collectionWithSingleNull = new ArrayList<>();
+        collectionWithSingleNull.add(null);
+        assertThat(testObject.myCollectionWithSingleNullValue, is(collectionWithSingleNull));
+        final Collection<String> collectionWithNullValues = new ArrayList<>();
+        collectionWithNullValues.add(null);
+        collectionWithNullValues.add("string");
+        collectionWithNullValues.add(null);
+        collectionWithNullValues.add("null");
```
Comment (ndtaylor) [inline demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java]: You can use `CollectionUtils.newArrayList()`.

File: demo/src/test/java/com/workday/autoparse/json/demo/InstanceUpdaterTest.java [hunk 3]
```diff
@@ -353,6 +358,19 @@ public void testJSONArray() throws Exception {
...
+    @Test
+    public void testCollectionWithNullValueUpdate() throws Exception {
+        TestObject testObject = new TestObject();
+        testObject.myCollectionWithNullValues = Arrays.asList("test", null, "null");;
+
+        Map<String, Object> updates = getUpdateMapFromFile("update-json-array-with-nulls.json");
+        
+        TestObject$$JsonObjectParser.INSTANCE.updateInstanceFromMap(testObject, updates, CONTEXT);
+
+        Collection<String> expected = Arrays.asList("one", null, "null");
```
Comment (ndtaylor) [inline demo/src/test/java/com/workday/autoparse/json/demo/InstanceUpdaterTest.java]: Use `CollectionUtils.newArrayList()`. Same above.

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java [hunk 4]
```diff
@@ -254,6 +260,15 @@ private void testParse(String fileName) throws Exception {
...
+        assertThat(testObject.myCollectionWithNullValues, is(collectionWithNullValues));
```
Comment (ndtaylor) [inline demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java]: Use `assertEquals`.

File: core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java [hunk 5]
```diff
@@ -929,6 +931,9 @@ public static JSONArray parseAsJsonArray(JsonReader reader, String key) throws I
                 @SuppressWarnings("unchecked")
                 T castItem = (T) o;
                 parsedItem = castItem;
+            } else if (o == null) {
+                collection.add(null);
+                continue;
```
Comment (ndtaylor) [inline core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java]: Move this check to the end of the `else if` chain. I would also add a comment ("The parsed array has an explicit null, so add a null to the collection"), since it's a little confusing that `parsedItem` being null results in an exception, but adding null directly to the collection is okay.

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java [hunk 6]
```diff
@@ -254,6 +260,15 @@ private void testParse(String fileName) throws Exception {
         assertNull("testObject.myNullCollection", testObject.myNullCollection);
         assertEquals("testObject.myDefaultCollection", Collections.singleton("the one"),
                      testObject.myDefaultCollection);
+        final Collection<String> collectionWithSingleNull = new ArrayList<>();
+        collectionWithSingleNull.add(null);
+        assertThat(testObject.myCollectionWithSingleNullValue, is(collectionWithSingleNull));
```
Comment (ndtaylor) [inline demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java]: Use `assertEquals`. Also, I think you could just use `Collections.singleton(null)` without the intermediate variable.

File: demo/src/test/resources/com/workday/autoparse/json/demo/update-json-array-with-nulls.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+  "object": "update",
+  "myCollectionWithNullValues": [
+    "one",
+    null,
+    "null"
+  ]
+}
```
Comment (ndtaylor) [inline demo/src/test/resources/com/workday/autoparse/json/demo/update-json-array-with-nulls.json]: Add a newline.

File: demo/src/test/java/com/workday/autoparse/json/demo/InstanceUpdaterTest.java [hunk 8]
```diff
@@ -26,6 +30,7 @@
 import static com.workday.autoparse.json.demo.InstanceUpdaterTestUtils.CONTEXT;
 import static com.workday.autoparse.json.demo.InstanceUpdaterTestUtils.getUpdateMapFromFile;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.hamcrest.CoreMatchers.is;
```
Comment (ndtaylor) [inline demo/src/test/java/com/workday/autoparse/json/demo/InstanceUpdaterTest.java]: I think all the changes to the imports should go away when you use `assertEquals`. Same in the other test file.

**Review by ndtaylor — CHANGES_REQUESTED** [review]
Looks good overall. Just a few minor comments.

**Comment by ndtaylor** [comment]
> No, should we add support for that possibility? Should probably be in another PR.

For consistency, it should probably be added. But I concur, it should be in a separate PR.

**Comment by ndtaylor** [comment]
Also, it doesn't look like the CircleCI build is running. It probably has something to do with the fact that I don't have admin rights anymore. Anyone want to take a look at that? :-)

File: core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java (hunk 5 above)
Comment (zsperske) [reply core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java]: Yeah that placing makes more sense, added the comment as well.

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java (hunk 2 above)
Comment (zsperske) [reply demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java]: Done.

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java (hunk 6 above)
Comment (zsperske) [reply demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java]: Singleton doesn't seem to work because you end up comparing a set to an array list. Should the type of the TestObject field change?

**Comment by zsperske** [comment]
> Also, it doesn't look like the CircleCI build is running. It probably has something to do with the fact that I don't have admin rights anymore. Anyone want to take a look at that? :-)

@traviswestbrook ?

File: core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java [hunk 9]
```diff
@@ -929,10 +931,13 @@ public static JSONArray parseAsJsonArray(JsonReader reader, String key) throws I
                 @SuppressWarnings("unchecked")
                 T castItem = (T) o;
                 parsedItem = castItem;
-            } else if (o instanceof JSONObject) {
+            }  else if (o instanceof JSONObject) {
```
Comment (ndtaylor) [inline core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java]: Remove the extra whitespace.

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java (hunk 6 above)
Comment (ndtaylor) [reply demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java]: Eh, probably just use CollectionUtils.newArrayList().

File: demo/src/test/java/com/workday/autoparse/json/demo/InstanceUpdaterTest.java [hunk 10]
```diff
@@ -353,6 +353,18 @@ public void testJSONArray() throws Exception {
         assertTrue("myJsonArray[2] instanceof JSONObject", testObject.myJsonArray.opt(2) instanceof JSONObject);
     }
 
+    @Test
+    public void testCollectionWithNullValueUpdate() throws Exception {
+        TestObject testObject = new TestObject();
+        testObject.myCollectionWithNullValues = Arrays.asList("test", null, "null");;
```
Comment (ndtaylor) [inline demo/src/test/java/com/workday/autoparse/json/demo/InstanceUpdaterTest.java]: `CollectionUtils.newArrayList()`. And you have one too many semicolons.

File: core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java [hunk 11]
```diff
@@ -929,10 +931,13 @@ public static JSONArray parseAsJsonArray(JsonReader reader, String key) throws I
...
                 parsedItem = convertJsonObject((JSONObject) o, typeClass, itemParser, context);
             } else if (o instanceof String && converter != null) {
                 parsedItem = converter.convert((String) o);
+            } else if (o == null) { //The parsed array has an explicit null, so add a null to the collection
```
Comment (ndtaylor) [inline core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java]: Put this on its own line.
```
} else if (o == null) {
   // comment
   collection.add(null);
   continue;
}
```
Also, grammar nazi says to put a period at the end of the commend and space after the `//`. :-)

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java [hunk 12]
```diff
@@ -254,6 +255,11 @@ private void testParse(String fileName) throws Exception {
         assertNull("testObject.myNullCollection", testObject.myNullCollection);
         assertEquals("testObject.myDefaultCollection", Collections.singleton("the one"),
                      testObject.myDefaultCollection);
+
+        Collection<String> expected = new ArrayList<>();
+        expected.add(null);
+        assertEquals(expected, testObject.myCollectionWithSingleNullValue);
```
Comment (ndtaylor) [inline demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java:261]: I would just do
```
assertEquals(CollectionUtils.newArrayList(null), testObject.myCollectionWithSingleNullValue);
```

File: demo/src/test/resources/com/workday/autoparse/json/demo/update-json-array-with-nulls.json (hunk 7 above)
Comment (ndtaylor) [reply demo/src/test/resources/com/workday/autoparse/json/demo/update-json-array-with-nulls.json]: Resolved

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java (hunk 12 above)
Comment (zsperske) [reply demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java:261]: CollectionUtils would fail its preconditions check if I did that I believe.

File: demo/src/test/java/com/workday/autoparse/json/demo/InstanceUpdaterTest.java (hunk 10 above)
Comment (zsperske) [reply demo/src/test/java/com/workday/autoparse/json/demo/InstanceUpdaterTest.java]: Done.

File: core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java (hunk 11 above)
Comment (zsperske) [reply core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java]: Done and done haha

File: core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java (hunk 9 above)
Comment (zsperske) [reply core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java]: Done.

**Comment by KennethNickles** [comment]
CI is now fixed. As soon as Nathan signs off on the PR I'll merge it.

**Review by traviswestbrook — APPROVED** [review]

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java (hunk 12 above)
Comment (ndtaylor) [reply demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java:261]: Oh, I forgot about that. Well, for future reference, you get around that by casting `null` to a `String`. 
```
CollectionUtils.newArrayList((String) null);
```
But you can leave it as is here.

**Review by ndtaylor — APPROVED** [review]

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java (hunk 12 above)
Comment (zsperske) [reply demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java:261]: Woah, that's interesting. Thanks for the tip, and for the review!
